File: harness/oracle/cases/model-context.ts
/**
 * Oracle case #30 (Wave 7 / B22): getModelContextLimit + consumers.
 *
 * Drives the REAL helpers:
 *   getModelContextLimit, hasExtendedContext, getSafeInputLimit
 *     (lib/llm/model-context-data.ts)
 *
 * Registry seam: the function reads plugin model-info, FALLBACK_PRICING, and the
 * registry default. We capture that injected context per provider ("providerctx"
 * rows) so the Rust port can be fed exactly what the real function saw, then the
 * "query" rows assert the three outputs.
 *
 * As of W4.7a this oracle INITIALIZES the real provider registry (the runtime way
 * — load the built plugins/dist bundles + `initializeProviderRegistry`), so
 * `getDefaultContextWindow(provider)` and `getProvider(provider).getModelInfo()`
 * return the real manifest values (e.g. ANTHROPIC default 200000, DEEPSEEK/Z_AI
 * 131072) instead of the empty-registry sentinel 8192 / empty model-info. The
 * Rust port's `get_model_context_limit` is fed exactly these injected values, so
 * the differential proves the leaf math; the registry values themselves are
 * cross-checked in `provider_registry_equivalence`. DEEPSEEK/Z_AI are included:
 * their fall-through models now resolve to 131072 via the registry default (step
 * 5), where an empty registry would have fallen to the hardcoded provider-default
 * table (no DEEPSEEK/Z_AI row → 8192).
 *
 * Run from inside the server checkout:
 *   cd ~/source/quilltap-server
 *   npx tsx ~/source/quilltap-v5/harness/oracle/cases/model-context.ts \
 *     > /tmp/oracle-model-context.ndjson
 */

import { createRequire } from 'node:module';
import { join } from 'node:path';
import {
  getModelContextLimit,
  hasExtendedContext,
  getSafeInputLimit,
} from '@/lib/llm/model-context-data';
import { FALLBACK_PRICING } from '@/lib/llm/pricing';
import {
  getProvider,
  getDefaultContextWindow,
  initializeProviderRegistry,
} from '@/lib/plugins/provider-registry';

const nodeRequire = createRequire(join(process.cwd(), 'noop.js'));
await initializeProviderRegistry(
  ['anthropic', 'openai', 'google', 'grok', 'deepseek', 'z-ai', 'openrouter', 'ollama', 'openai-compatible'].map((d) => {
    const m = nodeRequire(join(process.cwd(), 'plugins', 'dist', `qtap-plugin-${d}`, 'index.js'));
    return m.plugin || m.default?.plugin || m.default;
  }),
);

const rows: unknown[] = [];

const PROVIDERS = [
  'ANTHROPIC',
  'OPENAI',
  'GOOGLE',
  'GROK',
  'DEEPSEEK',
  'Z_AI',
  'OLLAMA',
  'OPENROUTER',
  'OPENAI_COMPATIBLE',
  'WEIRD_UNKNOWN', // exercises the final provider-default → 8192 fallback
];

for (const p of PROVIDERS) {
  const plugin: any = getProvider(p as any);
  const modelInfo =
    plugin && typeof plugin.getModelInfo === 'function'
      ? plugin.getModelInfo().map((m: any) => ({ id: m.id, contextWindow: m.contextWindow ?? null }))
      : [];
  const fp = ((FALLBACK_PRICING as any)[p] || []).map((m: any) => ({
    modelId: m.modelId,
    contextLength: m.contextLength ?? null,
  }));
  rows.push({
    kind: 'providerctx',
    provider: p,
    modelInfo,
    fallbackPricing: fp,
    registryDefault: getDefaultContextWindow(p as any),
  });
}

// (provider, model, maxResponseTokens) — covering every lookup branch.
const queries: Array<[string, string, number]> = [
  // exact overrides
  ['OPENAI', 'gpt-4-32k', 4096],
  ['OLLAMA', 'phi3:mini', 4096],
  ['OLLAMA', 'mistral:7b', 4096],
  ['OPENAI', 'gpt-3.5-turbo-16k', 4096],
  ['OPENAI', 'gpt-4-0613', 4096],
  ['OLLAMA', 'llama3.2:3b', 2048],
  // provider-prefixed overrides
  ['ANTHROPIC', 'claude-3-opus', 4096],
  ['OPENAI', 'gpt-4', 4096],
  ['GOOGLE', 'gemini-pro', 8192],
  // FALLBACK_PRICING hits (exact + substring)
  ['ANTHROPIC', 'claude-sonnet-4-5-20250929', 4096],
  ['OPENAI', 'gpt-4o', 4096],
  ['OPENAI', 'gpt-4o-mini', 4096],
  ['OPENAI', 'gpt-4o-2024-11-20', 4096],
  ['GROK', 'grok-2', 4096],
  // provider-default fallbacks (unknown model)
  ['ANTHROPIC', 'totally-unknown-xyz', 4096],
  ['OPENAI', 'totally-unknown-xyz', 4096],
  ['GOOGLE', 'totally-unknown-xyz', 4096],
  ['OLLAMA', 'totally-unknown-xyz', 4096],
  ['OPENROUTER', 'totally-unknown-xyz', 4096],
  ['OPENAI_COMPATIBLE', 'totally-unknown-xyz', 4096],
  // DEEPSEEK/Z_AI fall-through: the registry default (131072) now decides via
  // step 5, where an empty registry would have fallen to 8192 (no provider-default
  // table row). This is the W4.7a seam captured.
  ['DEEPSEEK', 'totally-unknown-xyz', 4096],
  ['Z_AI', 'totally-unknown-xyz', 4096],
  ['WEIRD_UNKNOWN', 'totally-unknown-xyz', 4096],
  // safe-input floor (huge response reserve clamps to 1000)
  ['OLLAMA', 'phi3:mini', 1000000],
  ['ANTHROPIC', 'claude-sonnet-4-5-20250929', 8192],
];
for (const [provider, model, maxResp] of queries) {
  rows.push({
    kind: 'query',
    provider,
    model,
    maxResponseTokens: maxResp,
    limit: getModelContextLimit(provider as any, model),
    extended: hasExtendedContext(provider as any, model),
    safeInput: getSafeInputLimit(provider as any, model, maxResp),
  });
}

for (const r of rows) process.stdout.write(JSON.stringify(r) + '\n');

File: crates/quilltap-core/src/model_context.rs
//! Port of v4's lib/llm/model-context-data.ts — the context-window lookup
//! (`getModelContextLimit`) and its two thin consumers (`hasExtendedContext`,
//! `getSafeInputLimit`).
//!
//! Registry seam: the lookup threads two registry calls (`getProvider(provider)
//! ?.getModelInfo()` and `getDefaultContextWindow(provider)`) and the pricing
//! module's `FALLBACK_PRICING` table through the middle of an otherwise pure
//! sequence. The function's own constant tables (`MODEL_CONTEXT_OVERRIDES`,
//! `DEFAULT_CONTEXT_BY_PROVIDER`) are ported as constants below; the
//! registry/pricing data is injected as parameters — the plugin model-info, the
//! provider's `FALLBACK_PRICING` rows, and the registry default. This mirrors v4
//! exactly while keeping the volatile data tables out of the core (the same
//! seam-injection pattern as `cheap_model` and `token_estimation`).

/// A plugin model-info entry, as far as the limit lookup reads it. `None`
/// context window is v4's `undefined`/falsy (skipped).
#[derive(Clone, Debug)]
pub struct ModelInfo {
    pub id: String,
    pub context_window: Option<i64>,
}

/// A `FALLBACK_PRICING` row, as far as the limit lookup reads it. `None` context
/// length is v4's `contextLength: null` (falsy → skipped).
#[derive(Clone, Debug)]
pub struct PricingRow {
    pub model_id: String,
    pub context_length: Option<i64>,
}

/// v4's `MODEL_CONTEXT_OVERRIDES` — per-model context windows that win over
/// everything else. Exact-key match (also used for the `provider/model` form).
fn model_context_override(key: &str) -> Option<i64> {
    let v = match key {
        // Ollama models
        "llama3.2:3b" => 131072,
        "llama3.1:8b" => 131072,
        "llama3.1:70b" => 131072,
        "mistral:7b" => 32768,
        "mixtral:8x7b" => 32768,
        "codellama:7b" => 16384,
        "phi3:mini" => 4096,
        "qwen2:7b" => 32768,
        // OpenRouter-specific models
        "anthropic/claude-3-opus" => 200000,
        "anthropic/claude-3-sonnet" => 200000,
        "anthropic/claude-3-haiku" => 200000,
        "openai/gpt-4-turbo" => 128000,
        "openai/gpt-4" => 8192,
        "google/gemini-pro" => 1000000,
        // Older OpenAI models
        "gpt-4-0613" => 8192,
        "gpt-4-32k" => 32768,
        "gpt-3.5-turbo-16k" => 16385,
        _ => return None,
    };
    Some(v)
}

/// v4's `DEFAULT_CONTEXT_BY_PROVIDER` — conservative per-provider fallback.
fn default_context_by_provider(provider: &str) -> Option<i64> {
    let v = match provider {
        "ANTHROPIC" => 200000,
        "OPENAI" => 128000,
        "GOOGLE" => 1000000,
        "GROK" => 131072,
        "OLLAMA" => 8192,
        "OPENROUTER" => 128000,
        "OPENAI_COMPATIBLE" => 8192,
        _ => return None,
    };
    Some(v)
}

/// v4's substring matcher: `m.id === name || m.id.includes(name) ||
/// name.includes(m.id)`. ASCII model ids, so `includes` is byte-substring.
fn id_matches(candidate: &str, name: &str) -> bool {
    candidate == name || candidate.contains(name) || name.contains(candidate)
}

/// Context-window size for a model, reproducing v4's `getModelContextLimit`
/// lookup order: exact override → `provider/model` override → plugin model-info
/// (when injected) → `FALLBACK_PRICING` → registry default (when it differs from
/// 8192) → the hardcoded provider default (else 8192).
///
/// `model_info` / `fallback_pricing` are the injected registry/pricing rows;
/// `registry_default` is `getDefaultContextWindow(provider)`. A zero/`None`
/// context value at any data stage is falsy in v4 and falls through, matching
/// the `if (modelInfo?.contextWindow)` / `if (modelPricing?.contextLength)`
/// guards.
pub fn get_model_context_limit(
    provider: &str,
    model_name: &str,
    model_info: &[ModelInfo],
    fallback_pricing: &[PricingRow],
    registry_default: i64,
) -> i64 {
    // 1. explicit override
    if let Some(v) = model_context_override(model_name) {
        return v;
    }
    // 2. provider-prefixed override (OpenRouter form)
    let prefixed = format!("{}/{}", provider.to_lowercase(), model_name);
    if let Some(v) = model_context_override(&prefixed) {
        return v;
    }
    // 3. plugin model-info (first match wins; falsy contextWindow falls through)
    if let Some(mi) = model_info.iter().find(|m| id_matches(&m.id, model_name)) {
        if let Some(cw) = mi.context_window {
            if cw != 0 {
                return cw;
            }
        }
    }
    // 4. fallback pricing (first match wins; null/0 contextLength falls through)
    if let Some(mp) = fallback_pricing
        .iter()
        .find(|m| id_matches(&m.model_id, model_name))
    {
        if let Some(cl) = mp.context_length {
            if cl != 0 {
                return cl;
            }
        }
    }
    // 5. registry default, only when it isn't the registry's own 8192 sentinel
    if registry_default != 8192 {
        return registry_default;
    }
    // 6. hardcoded provider default, else 8192
    default_context_by_provider(provider).unwrap_or(8192)
}

/// Resolve the effective context window — v4's `resolveContextWindow`
/// (`lib/llm/model-context-data.ts:154`) in the lookup form: the profile's
/// user-set Max Context wins, the name lookup is the fallback, and a zero or
/// negative column falls through rather than producing a zero-token budget.
///
/// v4 keeps both halves in one module; v5 splits them (see this module's header),
/// so the arithmetic half lives in [`crate::context_budget::resolve_context_window`]
/// and this is the composition with [`get_model_context_limit`]. Callers holding a
/// profile must route through one of the two rather than calling the lookup bare,
/// or the budget and the compression trigger end up on two different windows
/// (v4 bug 70).
pub fn resolve_context_window(
    provider: &str,
    model_name: &str,
    model_info: &[ModelInfo],
    fallback_pricing: &[PricingRow],
    registry_default: i64,
    profile_max_context: Option<i64>,
) -> i64 {
    // The profile wins before the lookup runs at all (v4 returns early).
    if let Some(c) = profile_max_context {
        if c > 0 {
            return c;
        }
    }
    get_model_context_limit(
        provider,
        model_name,
        model_info,
        fallback_pricing,
        registry_default,
    )
}

/// Whether the model supports extended (> 32k) context — v4's
/// `hasExtendedContext`. Deliberately NOT profile-aware: v4's `f933ba9c` left
/// this one on the bare lookup.
pub fn has_extended_context(
    provider: &str,
    model_name: &str,
    model_info: &[ModelInfo],
    fallback_pricing: &[PricingRow],
    registry_default: i64,
) -> bool {
    get_model_context_limit(
        provider,
        model_name,
        model_info,
        fallback_pricing,
        registry_default,
    ) > 32768
}

/// Safe input-context limit: the resolved window minus the response reserve and
/// a 10% safety buffer (`ceil`), floored at 1000 — v4's `getSafeInputLimit`.
///
/// Since `f933ba9c` v4 composes this from `computeSafeInputLimit` over
/// `resolveContextWindow` instead of re-deriving the arithmetic, so the formula
/// has exactly one owner ([`crate::context_budget::compute_safe_input_limit`])
/// and the profile's Max Context governs here too.
pub fn get_safe_input_limit(
    provider: &str,
    model_name: &str,
    model_info: &[ModelInfo],
    fallback_pricing: &[PricingRow],
    registry_default: i64,
    max_response_tokens: i64,
    profile_max_context: Option<i64>,
) -> i64 {
    let total = resolve_context_window(
        provider,
        model_name,
        model_info,
        fallback_pricing,
        registry_default,
        profile_max_context,
    );
    crate::context_budget::compute_safe_input_limit(total, max_response_tokens)
}

#[cfg(test)]
mod nanogpt_legacy_table_guard {
    use super::*;
    use crate::provider_manifest::Registry;

    /// P4.D101 — NanoGPT is deliberately ABSENT from v4's legacy per-provider
    /// fallback tables, and must stay absent.
    ///
    /// MEASURED at v4 `4cb1035e`: `NANOGPT` appears in v4's entire `lib/` tree
    /// exactly ONCE — `EmbeddingProfileProviderEnum` in
    /// `lib/schemas/common.types.ts`. It is in NONE of
    /// `DEFAULT_CONTEXT_BY_PROVIDER`, `LEGACY_RECOMMENDED_CHEAP_MODELS`,
    /// `PROVIDER_NAME_SUPPORT`, `PROVIDER_ATTACHMENT_CAPABILITIES`, or the
    /// pricing fallback data. Those tables are v4's PRE-plugin fallbacks; a
    /// plugin-era provider is served by the registry instead, and `Provider` is
    /// `z.string()` so no type forces an entry.
    ///
    /// Adding a NANOGPT row to any of them would be a v5-invented divergence.
    /// This test is the tripwire: it pins the absence AND proves the registry
    /// path still answers 131072, so the absence costs nothing.
    #[test]
    fn nanogpt_is_absent_from_the_legacy_table_but_served_by_the_registry() {
        assert_eq!(
            default_context_by_provider("NANOGPT"),
            None,
            "NANOGPT must NOT be in the legacy DEFAULT_CONTEXT_BY_PROVIDER table — \
             v4 has no such row (measured at 4cb1035e)"
        );

        // The manifest carries the real value, and the registry is consulted
        // first, so the legacy table is never reached for NanoGPT.
        let manifest = Registry::built_in().get_provider("NANOGPT").unwrap();
        assert_eq!(manifest.default_context_window, 131072);
        assert_eq!(
            resolve_context_window("NANOGPT", "openai/gpt-5-mini", &[], &[], 131072, None),
            131072
        );
    }
}
